File: benchmarking/locust/common/trace.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections.abc import Sequence
from locust import events
from locust.argument_parser import LocustArgumentParser
from locust.env import Environment
from opentelemetry import trace
from opentelemetry.context import Context
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace.sampling import SamplingResult, TraceIdRatioBased, Sampler
from opentelemetry.propagate import set_global_textmap
from opentelemetry.trace import Link, SpanKind, Tracer
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.trace.span import TraceState
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.util.types import Attributes
import logging

logger = logging.getLogger(__name__)

class UpdatableSampler(Sampler):
    def __init__(self, initial_probability: float = 0.0) -> None:
        self.sampler = TraceIdRatioBased(initial_probability)
        self.probability = initial_probability

    def update_probability(self, probability: float) -> None:
        self.sampler = TraceIdRatioBased(probability)
        self.probability = probability

    def should_sample(
        self,
        parent_context: Context | None,
        trace_id: int,
        name: str,
        kind: SpanKind | None = None,
        attributes: Attributes = None,
        links: Sequence[Link] | None = None,
        trace_state: TraceState | None = None,
    ) -> SamplingResult:
        return self.sampler.should_sample(
            parent_context, trace_id, name, kind, attributes, links, trace_state
        )

    def get_description(self) -> str:
        return f"UpdatableSampler(probability={self.probability})"

_initialized = False
_sampler = UpdatableSampler(0.0)


# All locust test files share one process-global TracerProvider, so the
# service name is fixed here rather than passed per-call.
SERVICE_NAME_VALUE = "substrate-locust"


def init_tracing() -> None:
    global _initialized, _sampler
    if _initialized:
        logger.info("Tracing already initialized, skipping.")
        return

    @events.init_command_line_parser.add_listener
    def _(parser: LocustArgumentParser) -> None:
        parser.add_argument(
            "--trace-probability",
            type=float,
            default=0.0,
            env_var="LOCUST_TRACE_PROBABILITY",
            help="Probability of tracing requests (0.0 to 1.0)",
            include_in_web_ui=True,
        )

    @events.init.add_listener
    def on_locust_init(environment: Environment, **kwargs) -> None:
        options = environment.parsed_options
        probability = getattr(options, 'trace_probability', 0.0)

        _sampler.update_probability(probability)

        resource = Resource(attributes={
            SERVICE_NAME: SERVICE_NAME_VALUE
        })
        provider = TracerProvider(sampler=_sampler, resource=resource)

        # Always add the exporter so it's ready if probability is increased later
        processor = BatchSpanProcessor(OTLPSpanExporter())
        provider.add_span_processor(processor)

        trace.set_tracer_provider(provider)
        set_global_textmap(TraceContextTextMapPropagator())
        logger.info(f"Tracing initialized for {SERVICE_NAME_VALUE} with initial probability {probability}")

    @events.test_start.add_listener
    def on_test_start(environment: Environment, **kwargs) -> None:
        options = environment.parsed_options
        probability = getattr(options, 'trace_probability', 0.0)
        _sampler.update_probability(probability)
        logger.info(f"Test started, updated trace probability to {probability}")

    _initialized = True


def get_tracer(name: str) -> Tracer:
    return trace.get_tracer(name)

File: internal/benchmarking/boomer/trace/trace.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package trace wires the OTLP tracer + W3C propagator used by the boomer
// load-test workers. Same shape as internal/serverboot.InitTracing but
// without the metrics/HTTP surface that workers don't need.
package trace

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"sync/atomic"

	"github.com/google/uuid"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.40.0"
)

// UpdatableSampler wraps a TraceIDRatioBased sampler whose probability can
// be swapped at runtime (mirrors common/trace.py:UpdatableSampler on the
// Python side). The OTel TracerProvider only accepts one Sampler at
// construction, so to change probability later you have to hand it a
// stable wrapper that internally swaps.
type UpdatableSampler struct {
	inner atomic.Pointer[sdktrace.Sampler]
}

func NewUpdatableSampler(initialProbability float64) *UpdatableSampler {
	s := &UpdatableSampler{}
	s.UpdateProbability(initialProbability)
	return s
}

func (u *UpdatableSampler) UpdateProbability(p float64) {
	s := sdktrace.ParentBased(sdktrace.TraceIDRatioBased(p))
	u.inner.Store(&s)
}

func (u *UpdatableSampler) ShouldSample(p sdktrace.SamplingParameters) sdktrace.SamplingResult {
	return (*u.inner.Load()).ShouldSample(p)
}

func (u *UpdatableSampler) Description() string {
	return "UpdatableSampler"
}

// Init registers a global TracerProvider with the given sampler and the
// W3C TraceContext propagator. The OTLP gRPC exporter honors
// OTEL_EXPORTER_OTLP_ENDPOINT.
func Init(ctx context.Context, serviceName string, sampler sdktrace.Sampler) (*sdktrace.TracerProvider, error) {
	if serviceName == "" {
		return nil, fmt.Errorf("serviceName is required")
	}
	res, err := resource.New(ctx,
		resource.WithTelemetrySDK(),
		resource.WithSchemaURL(semconv.SchemaURL),
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.ServiceInstanceID(uuid.NewString()),
		),
		resource.WithFromEnv(),
	)
	// Partial-resource warnings come back as errors; tolerate them the way
	// serverboot.InitTracing does so a missing OTEL_* env var doesn't fail
	// the whole worker.
	if errors.Is(err, resource.ErrPartialResource) || errors.Is(err, resource.ErrSchemaURLConflict) {
		slog.WarnContext(ctx, "partial telemetry resource", slog.Any("err", err))
	} else if err != nil {
		return nil, fmt.Errorf("create resource: %w", err)
	}

	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithInsecure())
	if err != nil {
		return nil, fmt.Errorf("create OTLP exporter: %w", err)
	}

	// Route OTel SDK errors (export failures, queue drops) into slog so they
	// land in boomer-worker's stdout — runner.py pumps that into logs.txt.
	// Without this they go to the SDK's default handler (stderr via log.Println)
	// and can be lost depending on stream wiring.
	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
		slog.Error("otel error", slog.String("err", err.Error()))
	}))

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sampler),
		sdktrace.WithBatcher(exporter),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.TraceContext{})
	return tp, nil
}
